Scene graph implementation

zmughal opened the pull request:
Connects with: <https://github.com/Intertangle/p5-Renard-Jacquard/pull/4>, <https://github.com/Intertangle/p5-Renard-Punchcard/pull/2>.

Steps for prototype:
- [x] Draw PDF pages to SVG.
- [x] Draw PDF pages as grid to SVG.
- [x] Draw PDF pages with box layout (margins around pages) to SVG.
- [x] Draw PDF pages as `Grid[Grid[Box[Page]]]` layout to SVG.
- [x] Draw PDF pages to Cairo surface (with Cairo surface size as big as all the pages).
- [x] Show which pages are visible in view-port of Cairo surface.
- [x] Draw PDF pages to Cairo surface (with Cairo surface the size of viewport).
- [x] Show page under cursor on hover.
- [x] Change cursor to [I-beam pointer](https://en.wikipedia.org/wiki/Cursor_(user_interface)#I-beam_pointer) when over text.
<del>

- [ ] Select text via mouse drag under cursor for single page.
- [ ] Select text via mouse drag under cursor across page boundaries.
- [ ] Implement changing № of columns for grid displayed on Cairo surface.
- [ ] Implement continuous / non-continuous view while keeping previous features.
- [ ] Implement zooming while keeping previous features.

</del>

zmughal committed:
Start grid test and PageActor

zmughal committed:
Add dep on Renard::Jacquard

zmughal committed:
Draw grid of PDF pages rendered as PNGs in SVG

zmughal committed:
Move size and bounds calculations to roles

zmughal committed:
Implement Grid[Grid[Box[Page]]] scene-graph

zmughal committed:
Time grid rendering

zmughal reviewed (commented):
24ceee80d5a3c985b3ce9cd4f0d456258f15cd1f
```
git show 'scenegraph-jacquard^{/Draw grid of PDF pages rendered as PNGs in SVG}'
```
- [X] Draw PDF pages to SVG.

![00-draw-pdf-pages-to-svg](https://user-images.githubusercontent.com/94489/89657624-fb3b1700-d892-11ea-8a35-2e7f0e0c9533.png)

zmughal reviewed (commented):
24ceee80d5a3c985b3ce9cd4f0d456258f15cd1f
```
git show 'scenegraph-jacquard^{/Draw grid of PDF pages rendered as PNGs in SVG}'
```
- [X] Draw PDF pages as grid to SVG.
![01-draw-pdf-pages-as-grid-to-svg](https://user-images.githubusercontent.com/94489/89657857-5ff67180-d893-11ea-895c-4b30a49305b2.png)

zmughal reviewed (commented):
6cdebb4bb00f684b2c554f49f01d3cec4291dc44
```
git show 'scenegraph-jacquard^{/Implement Grid.Grid.Box.Page.* scene-graph}'
```
- [X] Draw PDF pages with box layout (margins around pages) to SVG.
![02-draw-pdf-pages-with-box-layout-margins-to-svg](https://user-images.githubusercontent.com/94489/89657977-9207d380-d893-11ea-8569-ad51f6d9e5d1.png)

zmughal reviewed (commented):
6cdebb4bb00f684b2c554f49f01d3cec4291dc44
```
git show 'scenegraph-jacquard^{/Implement Grid.Grid.Box.Page.* scene-graph}'
```
- [X] Draw PDF pages as `Grid[Grid[Box[Page]]]` layout to SVG.
![03-draw-pdf-pages-as-grid-grid-box-page-layout-to-svg](https://user-images.githubusercontent.com/94489/89658080-b2d02900-d893-11ea-8f91-575e83c91c07.png)

zmughal committed:
Split SVG generation to functions

zmughal committed:
Use branch p5-Renard-Taffeta@cairo-transform-matrix

zmughal committed:
Draw PDF pages to Cairo surface (set size to bounds)

With Cairo surface size set to the bounds of the scene graph.

zmughal reviewed (commented):
192542e39313e4aa9e318599542a19b3a6f72c9b

```
git show 'scenegraph-jacquard^{/Draw PDF pages to Cairo surface \(set size to bounds\)}'
```
- [X] Draw PDF pages to Cairo surface (with Cairo surface size as big as all the pages).

![t-grid pl (1)_816](https://user-images.githubusercontent.com/94489/90494907-9d67c400-e109-11ea-9617-e6ed27ee1801.png)
![t-grid pl (1)_819](https://user-images.githubusercontent.com/94489/90494916-a062b480-e109-11ea-9fcd-44972078d08f.png)

zmughal committed:
Add status bar to window

zmughal committed:
Add bounds to PageActor

zmughal committed:
Update status bar with pages in viewport

zmughal reviewed (commented):
d2a80d8985f8574e6dc1a64ea36c459580beb48a

```
git show 'scenegraph-jacquard^{/Update status bar with pages in viewport}'
```

- [X] Show which pages are visible in view-port of Cairo surface.


![curie-t-grid-statusbar-update-2020-08-29_19 27 00](https://user-images.githubusercontent.com/94489/91648661-ae7cd300-ea2f-11ea-85f8-f24979d2c5b2.gif)

zmughal committed:
Only do translation matrix for Layout nodes

zmughal committed:
Store bounds for highlighting

zmughal reviewed (commented):
52ecde90614cae859feef96166c81d7253b3d49c

```
git show 'scenegraph-jacquard^{/Store bounds for highlighting}'
```

```
T_GRID_HIGHLIGHT_BOUNDS=1 perl -MCarp::Always ./t-grid.pl
```


Highlighted bounds.

![t-grid pl (1)_831](https://user-images.githubusercontent.com/94489/91649636-8bf0b700-ea3b-11ea-958e-bebe1d38f003.png)

zmughal committed:
Mark when done updating layouts

zmughal committed:
Create widget that handles scrolling and visibility

The widget extends Gtk3::DrawingArea and implements Gtk3::Scrollable. It
only draws pages that are visible while keeping the size of the drawing
area to that of the Gtk3::ScrolledWindow.

zmughal reviewed (commented):
b32d15081358ecb2d20e1e65ab9826e9e72fa9e5

```
git show 'scenegraph-jacquard^{/Create widget that handles scrolling and visibility}'
```

- [X] Draw PDF pages to Cairo surface (with Cairo surface the size of viewport).

![t-grid pl (1)_832](https://user-images.githubusercontent.com/94489/91672310-76d75f00-eaf3-11ea-8c15-b26815fc7008.png)

zmughal committed:
Pages under mouse cursor using motion-notify-event

zmughal reviewed (commented):
c7f3ec22eb86708945e656ff416889d15716a36f

```
git show 'scenegraph-jacquard^{/Pages under mouse cursor using motion-notify-event}'
```

- [X] Show page under cursor on hover.

![curie-t-grid-page-tooltip-2020-08-30_20 05 20](https://user-images.githubusercontent.com/94489/91673963-4a741080-eafc-11ea-8a8e-45117b1cbb98.gif)

mohawk2 commented:
I got an email from AppVeyor saying a build on this PR had failed. Any idea why it's emailing me? :-)

zmughal commented:
@mohawk2, because AppVeyor's notification settings are terribly confusing. I'm going to try to restrict them again.

I have several PDL team members under a PDLPorters role on my account. I think that the easiest approach is to use <https://www.appveyor.com/docs/team-setup/#deny-all-approach> and then go through all the PDL repos and enable the PDLPorters all the access.

zmughal commented:
@mohawk2 , it should be fixed now via <https://github.com/orbital-transfer/p5-Orbital-Payload-Service-AppVeyor/commit/c15d8c659573b319f4c449da2e209a00597fc211>. Permissions for all repos on my AppVeyor account are now restricted to the associated role for that repo's GitHub organization. Let me know if you get any extra build notifications.

zmughal committed:
Use ArrayRef[HashRef] for views instead of parallel ArrayRef's

mohawk2 commented:
Will do! Wasn't bothered as such, but thought you'd want to know.

zmughal committed:
Fix scale

zmughal committed:
Find text block under mouse cursor

- Highlight the bounding box of the block.
- Add a selectable label which contains the text of the block.

zmughal reviewed (commented):
eea4ed802c26dae5789158cce9d989f6c4c76648

```
/Find text block under mouse cursor 
```

![curie-t-grid-highlight-text-2020-09-01_21 55 40](https://user-images.githubusercontent.com/94489/91927077-4aac1180-ec9e-11ea-9389-330ebd1109f4.gif)

zmughal committed:
Highlight boxes for multiple levels

Uses levels of (block, line, char).

zmughal reviewed (commented):
1891b46aae55a0e8070b75e806a0e072df31ed17

```
/Highlight boxes for multiple levels
```

![curie-t-grid-highlight-text-layers-2020-09-01_22 47 14](https://user-images.githubusercontent.com/94489/91931983-92d13100-ecaa-11ea-93ee-2f679c6cfca7.gif)

zmughal committed:
Change cursor to I-beam pointer when over text

Only when over char level.

zmughal reviewed (commented):
b9feb904bdd5720fdf3fd598767a408a7b448649

```
/Change cursor to I-beam pointer when over text
```


- [X] Change cursor to [I-beam pointer](https://en.wikipedia.org/wiki/Cursor_(user_interface)#I-beam_pointer) when over text.

![curie-t-grid-change-cursor-2020-09-01_23 42 20](https://user-images.githubusercontent.com/94489/91932992-784c8700-ecad-11ea-9cd6-f2b555e3c276.gif)
